bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,507 @@
+
+    private final Set<String> allHandles = new HashSet<>(); // must be initialized
+
+    // gets replaced if Store becomes available. storageFacade handles all the times when
+    // the actual storage becomes unavilable and locking issues.
+    private StorageFacade storageFacade = new StorageFacade(null);
+
+    private final Set<StorageCipher> allAvailableStorageCiphers = new LinkedHashSet<>();
+    private Optional<@Nullable StorageCipher> storageCipher = Optional.empty();
+
+    private final Logger logger = LoggerFactory.getLogger(OAuthStoreHandlerImpl.class);
+
+    @SuppressWarnings("null") // Suppress for @NonNull Optional<@NonNull StorageCipher>
+    @Activate
+    public void activate(Map<String, Object> properties) throws GeneralSecurityException {
+        // this allows future implementations to change cipher by just setting the CIPHER_TARGET
+        String cipherTarget = (String) properties.getOrDefault("CIPHER_TARGET", SymmetricKeyCipher.CIPHER_ID);
+
+        // choose the cipher by the cipherTarget
+        storageCipher = allAvailableStorageCiphers.stream()
+                .filter(cipher -> cipher.getUniqueCipherId().equals(cipherTarget)).findFirst();
+
+        logger.debug("Using Cipher: {}", storageCipher
+                .orElseThrow(() -> new GeneralSecurityException("No StorageCipher with target=" + cipherTarget)));
+    }
+
+    /**
+     * Deactivate and free resources. Never set the storageFacade to null
+     * as this will cause NullPointerExceptions. StorageFacade can internally
+     * handle removed storage.
+     */
+    @Deactivate
+    public void deactivate() {
+        storageFacade.close(); // this removes old entries
+        // DS will take care of other references
+    }
+
+    @Override
+    public AccessTokenResponse loadAccessTokenResponse(@NonNull String handle) throws GeneralSecurityException {
+        AccessTokenResponse accessTokenResponseFromStore = (AccessTokenResponse) storageFacade.get(handle,
+                ACCESS_TOKEN_RESPONSE);
+
+        if (accessTokenResponseFromStore == null) {
+            // token does not exist
+            return null;
+        }
+
+        AccessTokenResponse decryptedAccessToken = decryptToken(accessTokenResponseFromStore);
+        return decryptedAccessToken;
+    }
+
+    @Override
+    public void saveAccessTokenResponse(@NonNull String handle, @Nullable AccessTokenResponse pAccessTokenResponse) {
+        AccessTokenResponse accessTokenResponse = pAccessTokenResponse;
+        String storageKeyAccessToken = ACCESS_TOKEN_RESPONSE.getKey(handle);
+        if (accessTokenResponse == null) {
+            accessTokenResponse = new AccessTokenResponse(); // put empty
+        }
+
+        AccessTokenResponse encryptedToken;
+        try {
+            encryptedToken = encryptToken(accessTokenResponse);
+        } catch (GeneralSecurityException e) {
+            logger.error("Unable to encrypt token, storing as-is", e);
```

Comment:
As you are doing a fallback logic, wouldn't a WARN be enough?

Reply:
Changed to warn.  I've a feeling this should be thrown instead.